### PR TITLE
Empty Blazor Server Project Template

### LINKED ISSUE #32832: Empty Blazor project template
The default Blazor template has a bunch of stuff in it to help you get started. This means you have a bunch of stuff to delete once you want to build your own app. We should have a way to build an empty Blazor project that has only minimum of what you need. This could potentially be an option on the existing templates.

### PR DESCRIPTION
**Pull request by cerasamson**
# Empty Blazor Server Project Template
Created an empty Blazor server template from the existing one. All contents in the Data folder were removed, along with additional HTML/CSS formatting. Authentication is removed entirely.

## Screenshot of template:
image

Fixes #32832

### CODE REVIEW

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/_Imports.razor [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿@using Microsoft.AspNetCore.Components.Routing
+@using Microsoft.AspNetCore.Components.Web
+@using Microsoft.JSInterop
+@using EmptyBlazorServerWeb_CSharp
+@using EmptyBlazorServerWeb_CSharp.Shared
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/_Imports.razor:5]:
```suggestion
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/wwwroot/css/site.css [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    padding: 1rem 1rem 1rem 3.7rem;
+    color: white;
+}
+
+.blazor-error-boundary::after {
+    content: "An error has occurred."
+}
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/wwwroot/css/site.css:28]: @cerasamson have you had a chance to test out this bit to ensure the error boundary / error ui continues to work with this subset of styling?

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/ide.host.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "$schema": "http://json.schemastore.org/vs-2017.3.host",
+  "order": 600,
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/ide.host.json:3]: Perhaps this should be a lower "order" than the regular templates (ex. `650`?)

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/Program.Main.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+using Microsoft.AspNetCore.Components;
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/Program.Main.cs:1]: May be able to remove this file entirely depending on the above discussion

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  "thirdPartyNotices": "https://aka.ms/aspnetcore/7.0-third-party-notices",
+  "tags": {
+    "language": "C#",
+    "type": "project"
+  },
+  "sourceName": "EmptyBlazorServerWeb-CSharp",
+  "preferNameDirectory": true,
+  "guids": [
+    "09732173-2cef-46b7-83db-1334bcb079d3", // Tenant ID
+    "53bc9b9d-9d6a-45d4-8429-2a2761773502" // Client ID
```
[the remark this answers is not in the document]
Comment (Eilon) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:25]: These are specified for this purpose: https://github.com/dotnet/templating/wiki/Reference-for-template.json#guids

TL;DR: This is a list of GUIDs that are used in the template, but should be auto-replaced when someone uses the template (`dotnet new ...`), and this makes sure that they are consistently replaced throughout the template.

So:
1. If these GUIDs are still used in this template and should be dynamically generated, keep this here
2. If these GUIDs are still used in this template but should not be dynamically generated, remove this (this is unlikely)
3. If these GUIDs are not used in this template, remove this

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    {
+      "id": "restore",
+      "condition": "(!skipRestore)",
+      "description": "Restore NuGet packages required by this project.",
+      "manualInstructions": [
+        {
+          "text": "Run 'dotnet restore'"
+        }
+      ],
+      "actionId": "210D431B-A78B-4D2F-B762-4ED3E3EA9025",
```
[the remark this answers is not in the document]
Comment (Eilon) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:204]: These action ids are "well-known GUIDs," so if you want to use this action, that's the GUID to use. It is not related at all to the earlier GUID list.

[the remark this answers is not in the document]
Comment (Eilon) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:204]: For this particular GUID, more info here: https://github.com/dotnet/templating/wiki/Post-Action-Registry#restore-nuget-packages

**Comment by TanayParikh** [comment]
Screen Shot 2022-06-07 at 12 09 55 PM

Does anyone know if we need to do something special for VS4Mac to pickup these templates? @mrward?

**Comment by Eilon** [comment]
Some questions:

1. Is there any "list of all the templates" that this needs to be added to, or is it magically included in the build and will be included in VS?
2. Will this template be included in whatever test(s) we have for templates, to make sure they get created and compile/run correctly?

**Comment by mrward** [comment]
@TanayParikh - VS Mac will need a change to show this project template. Will this only be in the .NET 7 SDK templates and not backported to the .NET 6 SDK?

**Comment by TanayParikh** [comment]
> Will this only be in the .NET 7 SDK templates and not backported to the .NET 6 SDK?

That's correct.

**Comment by TanayParikh** [comment]
> Is there any "list of all the templates" that this needs to be added to, or is it magically included in the build and will be included in VS?

From my understanding the current changeset should be sufficient. @cerasamson verified this new template shows up with `dotnet new --list` as well as the VS new project UI.

> Will this template be included in whatever test(s) we have for templates, to make sure they get created and compile/run correctly? (Again, this might be magical and automatic, but I don't know.)

We'll need to update/add the test for coverage in this area (at the very least update the test which validates the appropriate files show up with `dotnet new blazorserverempty`).

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (TanayParikh) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:25]: @cerasamson are these still used elsewhere in the template, or could we potentially remove them?

[the remark this answers is not in the document]
Comment (cerasamson) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:25]: I believe these can be removed. Since authentication was removed, I do not think the GUIDs are used for anything else and are no longer being generated anywhere in the template.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        },
+        {
+          "condition": "(UseProgramMain)",
+          "exclude": [
+            "Program.cs"
+          ],
+          "rename": {
+            "Program.Main.cs": "Program.cs"
+          }
+        },
```
[the remark this answers is not in the document]
Comment (cerasamson) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json]: You are right. I removed these modifiers as well as Program.Main.cs.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/localize/templatestrings.cs.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{
+  "author": "Microsoft",
+  "name": "Blazor Server App Empty",
+  "description": "An empty project template for creating a Blazor server app that runs server-side inside an ASP.NET Core app and handles user interactions over a SignalR connection. This template does not have any content in it.",
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/localize/templatestrings.cs.json:4]: This updated description will need to be reflected in the other template files.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/localize/templatestrings.cs.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  "symbols/iisHttpPort/description": "Port number to use for the IIS Express HTTP endpoint in launchSettings.json.",
+  "symbols/iisHttpsPort/description": "Port number to use for the IIS Express HTTPS endpoint in launchSettings.json. This option is only applicable when the parameter no-https is not used (no-https will be ignored if either IndividualAuth or OrganizationalAuth is used).",
+  "symbols/Framework/description": "The target framework for the project.",
+  "symbols/Framework/choices/net7.0/description": "Target net7.0",
+  "symbols/skipRestore/description": "If specified, skips the automatic restore of the project on create.",
+  "symbols/UseProgramMain/displayName": "Use a Program class with a Main method",
+  "symbols/UseProgramMain/description": "Whether to generate an explicit Program class and Main method instead of top-level statements.",
+  "postActions/restore/description": "Restore NuGet packages required by this project.",
+  "postActions/restore/manualInstructions/default/text": "Run 'dotnet restore'"
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/localize/templatestrings.cs.json:16]: Do we still need all of these, I believe we may've removed some of these options.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+      "description": "If specified, skips the automatic restore of the project on create.",
+      "defaultValue": "false"
+    },
+    "UseProgramMain": {
+      "type": "parameter",
+      "datatype": "bool",
+      "defaultValue": "false",
+      "displayName": "Do not use top-level statements",
+      "description": "Whether to generate an explicit Program class and Main method instead of top-level statements."
+    }
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:167]: Given we removed the `modifier` for this, I believe we can remove this symbol as well. Are some of these other symbols potentially no longer necessary as well?

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    "Blazor",
+    "Empty"
+  ],
+  "name": "Blazor Server App Empty",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An empty project template for creating a Blazor server app that runs server-side inside an ASP.NET Core app and handles user interactions over a SignalR connection. This template does not have any content in it.",
+  "groupIdentity": "Microsoft.Web.Blazor.Server.Empty",
+  "precedence": "9000",
+  "identity": "Microsoft.Web.Blazor.Server.Empty.CSharp.7.0",
+  "shortName": "emptyblazorserver",
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:15]: I believe @danroth27 prefers: `blazorserver-empty`

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/Program.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+#if (RequiresHttps)
+    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+#else
+}
+
+#endif
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/Program.cs:21]: I believe this may actually lead to different behavior than our current blazorserver template:

https://github.com/dotnet/aspnetcore/blob/0ddda008079da4bdb81ff9bdbd9621dec89e0a53/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Program.cs#L123-L144


```suggestion
#if (RequiresHttps)
if (!app.Environment.IsDevelopment())
{
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();
#endif
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/dotnetcli.host.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:1]: Have you had a chance to test out these CLI options?

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/ide.host.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      "name": "HttpsPort",
+      "useHttps": true
+    }
+  ],
+  "symbolInfo": [
+    {
+      "id": "UseProgramMain",
+      "isVisible": true
+    }
+  ],
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/ide.host.json:27]: Can this be removed entirely?

```suggestion
```

File: src/ProjectTemplates/Web.ProjectTemplates/EmptyBlazorServerWeb-CSharp.csproj.in [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>${DefaultNetCoreTargetFramework}</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <NoDefaultLaunchSettingsFile Condition="'$(ExcludeLaunchSettings)' == 'True'">True</NoDefaultLaunchSettingsFile>
+    <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">BlazorServerWeb_CSharp</RootNamespace>
```
Comment (TanayParikh) [inline src/ProjectTemplates/Web.ProjectTemplates/EmptyBlazorServerWeb-CSharp.csproj.in:8]: Should this be the following? 

```suggestion
    <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">EmptyBlazorServerWeb_CSharp</RootNamespace>
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/dotnetcli.host.json (hunk 13 above)
Comment (cerasamson) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:1]: Yes, these CLI options are working.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/wwwroot/css/site.css (hunk 2 above)
Comment (cerasamson) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/wwwroot/css/site.css]: Yes, both elements are working.

**Comment by japui** [comment]
Modelo de Projeto de Servidor Blazor Vazio

File: src/ProjectTemplates/test/EmptyBlazorServerTemplateTest.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Testing;
+using Templates.Test.Helpers;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Templates.Test;
+
+public class EmptyBlazorServerTemplateTest : BlazorTemplateTest
```
Comment (TanayParikh) [inline src/ProjectTemplates/test/EmptyBlazorServerTemplateTest.cs:17]: This isn't running in the CI as there was some refactoring of the project template test classes recently. I believe you'll have to move this over into [here](https://github.com/dotnet/aspnetcore/tree/db608a5f8db4de960af55961c41f3e22a776d7cd/src/ProjectTemplates/test/Templates.Blazor.Tests)

https://github.com/dotnet/aspnetcore/tree/db608a5f8db4de960af55961c41f3e22a776d7cd/src/ProjectTemplates/test

File: src/ProjectTemplates/Shared/Project.cs [hunk 17]
```diff
@@ -466,7 +466,7 @@ private void CaptureBinLogOnFailure(ProcessEx result)
         if (result.ExitCode != 0 && !string.IsNullOrEmpty(ArtifactsLogDir))
         {
             var sourceFile = Path.Combine(TemplateOutputDir, "msbuild.binlog");
-            Assert.True(File.Exists(sourceFile), $"Log for '{ProjectName}' not found in '{sourceFile}'.");
+            Assert.True(File.Exists(sourceFile), $"Log for '{ProjectName}' not found in '{sourceFile}'. Execution output:{result.Output}~Error:~{result.Error}~{result}");
```
Comment (TanayParikh) [inline src/ProjectTemplates/Shared/Project.cs:469]: I'm tempted to keep this part of the logs long term, I feel like it could assist with debugging.

For context; the current output with this is:

```
[1.501s] [Templates.Blazor.Test.EmptyBlazorServerTemplateTest] [Error] Test threw an exception.
Xunit.Sdk.TrueException: Log for 'AspNet.Rqm5pu3hoaxg' not found in '/datadisks/disk1/work/A35D08E8/w/A83309A0/e/Templates/BaseFolder/AspNet.Rqm5pu3hoaxg/msbuild.binlog'. Execution output:Microsoft (R) Build Engine version 17.3.0-preview-22256-03+72a076794 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1003: Specify a project or solution file. The current working directory does not contain a project or solution file.
~Error:~~Microsoft.AspNetCore.Internal.ProcessEx
```

We can probably change this to:

```suggestion
            Assert.True(File.Exists(sourceFile), $"Log for '{ProjectName}' not found in '{sourceFile}'. Execution output: {result.Output}");
```

File: src/ProjectTemplates/test/Templates.Tests/template-baselines.json [hunk 18]
```diff
@@ -1500,6 +1500,22 @@
...
+      "Files": [
+        "App.razor",
+        "appsettings.Development.json",
+        "appsettings.json",
+        "MainLayout.razor",
+        "Program.cs",
+        "_Imports.razor",
+        "Pages/Index.razor",
+        "Pages/_Host.cshtml",
+        "Properties/launchSettings.json",
```
Comment (TanayParikh) [inline src/ProjectTemplates/test/Templates.Tests/template-baselines.json:1515]: Can we just add a second test here with `exclude-launch-settings`?

**Review by TanayParikh — COMMENTED** [review]
One _last_ comment. Other than that, this looks great and ready for merging 😄

File: src/ProjectTemplates/test/Templates.Tests/template-baselines.json (hunk 18 above)
Comment (cerasamson) [reply src/ProjectTemplates/test/Templates.Tests/template-baselines.json:1515]: Yes, let me add that right now.

**Review by TanayParikh — APPROVED** [review]
Looks awesome, 🚢 it! 🎉

**Review by phenning — CHANGES_REQUESTED** [review]

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+      "condition": "(!skipRestore)",
+      "description": "Restore NuGet packages required by this project.",
+      "manualInstructions": [
+        {
+          "text": "Run 'dotnet restore'"
+        }
+      ],
+      "actionId": "210D431B-A78B-4D2F-B762-4ED3E3EA9025",
+      "continueOnError": true
+    }
```
[the remark this answers is not in the document]
Comment (cerasamson) [reply src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:196]: For empty templates, I don't think we want to open a project file by default, but @TanayParikh mentioned in the empty WASM PR that it might be a good option for `blazorwasm` and `blazorserver`.

File: src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+  ],
+  "name": "Blazor Server App Empty",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An empty project template for creating a Blazor server app that runs server-side inside an ASP.NET Core app and handles user interactions over a SignalR connection. This template does not have any content in it.",
+  "groupIdentity": "Microsoft.Web.Blazor.Server.Empty",
+  "precedence": "9000",
+  "identity": "Microsoft.Web.Blazor.Server.Empty.CSharp.7.0",
+  "shortName": "blazorserver-empty",
+  "thirdPartyNotices": "https://aka.ms/aspnetcore/7.0-third-party-notices",
+  "tags": [
```
Comment (phenning) [inline src/ProjectTemplates/Web.ProjectTemplates/content/EmptyBlazorServerWeb-CSharp/.template.config/template.json:17]: Sorry for the confusion, I meant that this section should be in ide.host.json. for clarity, keep this in the template.json

```
  "tags": {
    "language": "C#",
    "type": "project"
  },
```

This should go in the ide.host.json:
```
  "tags": [
    {
      "type": "projectType",
      "add": [ "Cloud", "Web" ],
      "remove": [ "*" ]
    },
    {
      "type": "platform",
      "add": [ "Linux", "macOS", "Windows" ]
    }
  ],
```

**Review by phenning — CHANGES_REQUESTED** [review]

**Review by phenning — APPROVED** [review]
Thanks!

**Comment by TanayParikh** [comment]
🎉 👏 🥇
